src/core/ucc_sbgp.c

```diff
@@ -0,0 +1,412 @@
+    local_ranks =
+        ucc_malloc(node_sbgp->group_size * sizeof(ucc_rank_t), "local_ranks");
+    if (!local_ranks) {
+        ucc_error("failed to allocate %zd bytes for local_ranks array",
+                  node_sbgp->group_size * sizeof(ucc_rank_t));
+        return UCC_ERR_NO_MEMORY;
+    }
+    for (i = 0; i < node_sbgp->group_size; i++) {
+        r = ucc_ep_map_eval(node_sbgp->map, i);
+        if (ucc_rank_on_local_socket(r, team)) {
+            local_ranks[sock_size] = r;
+            if (r == group_rank) {
+                sock_rank = sock_size;
+            }
+            sock_size++;
+        }
+    }
+    sbgp->group_size = sock_size;
+    sbgp->group_rank = sock_rank;
+    sbgp->rank_map   = local_ranks;
+    nlr_pos          = -1;
+    for (i = 0; i < sock_size; i++) {
+        if (nlr == local_ranks[i]) {
+            nlr_pos = i;
+            break;
+        }
+    }
+    if (nlr_pos > 0) {
+        if (sock_rank == 0)
+            sbgp->group_rank = nlr_pos;
+        if (sock_rank == nlr_pos)
+            sbgp->group_rank = 0;
+        SWAP(local_ranks[nlr_pos], local_ranks[0]);
+    }
+    if (sock_size > 1) {
+        sbgp->status = UCC_SBGP_ENABLED;
+    } else {
+        sbgp->status = UCC_SBGP_NOT_EXISTS;
+    }
+    return UCC_OK;
+}
+
+static inline ucc_status_t sbgp_create_node(ucc_team_topo_t *topo,
+                                            ucc_sbgp_t *     sbgp)
+{
+    ucc_team_t *team           = sbgp->team;
+    ucc_rank_t  group_size     = team->size;
+    ucc_rank_t  group_rank     = team->rank;
+    ucc_rank_t  max_local_size = 256;
+    ucc_rank_t  ctx_nlr        = topo->node_leader_rank_id;
+    ucc_rank_t  node_rank = 0, node_size = 0;
+    int         i;
+    ucc_rank_t *local_ranks;
+    local_ranks =
+        ucc_malloc(max_local_size * sizeof(ucc_rank_t), "local_ranks");
+    if (!local_ranks) {
+        ucc_error("failed to allocate %zd bytes for local_ranks array",
+                  max_local_size * sizeof(ucc_rank_t));
+        return UCC_ERR_NO_MEMORY;
+    }
+    for (i = 0; i < group_size; i++) {
+        if (ucc_rank_on_local_node(i, team)) {
+            if (node_size == max_local_size) {
+                max_local_size *= 2;
+                local_ranks = ucc_realloc(local_ranks,
+                                          max_local_size * sizeof(ucc_rank_t));
+                if (!local_ranks) {
+                    ucc_error(
+                        "failed to allocate %zd bytes for local_ranks array",
+                        max_local_size * sizeof(ucc_rank_t));
+                    return UCC_ERR_NO_MEMORY;
+                }
+            }
+            local_ranks[node_size] = i;
+
+            if (i == group_rank) {
+                node_rank = node_size;
+            }
+            node_size++;
+        }
+    }
+    if (0 == node_size) {
+        /* We should always have at least 1 local rank */
+        ucc_free(local_ranks);
```

Comment:
print warning/error

Reply:
same as above: this check is for linter. we always have at least "self" here.